projects/vdk-plugins/vdk-structlog/src/vdk/plugin/structlog/structlog_plugin.py

```diff
@@ -0,0 +1,171 @@
+
+Logging adapters a.k.a. bound loggers
+https://docs.python.org/3/howto/logging-cookbook.html#context-info
+
+Example:
+
+class BoundLogger(logging.LoggerAdapter):
+    def process(self, msg, kwargs):
+        # merge bound extra dict with existing exra dict if any
+        if "extra" in kwargs:
+            kwargs["extra"] = {**self.extra, **kwargs["extra"]}
+        else:
+            kwargs["extra"] = self.extra
+        return msg, kwargs
+
+
+log = logging.getLogger(__name__)
+bound = BoundLogger(log, {"bound_key": "bound_value"})
+bound.warning("From vdk_initialize", extra={"first": "first_value", "second": "second_value"})
+bound.warning("Something else from vdk_initialize")
+
+"""
+
+
+class StructlogPlugin:
+    @hookimpl(tryfirst=True)
+    def vdk_configure(self, config_builder: ConfigurationBuilder):
+        config_builder.add(
+            key=STRUCTLOG_LOGGING_METADATA_KEY,
+            default_value=",".join(list(STRUCTLOG_LOGGING_METADATA_DEFAULT.keys())),
+            description=(
+                f"Possible values: {STRUCTLOG_LOGGING_METADATA_ALL_KEYS}"
+                "User-defined key-value pairs added to the logger's context will be displayed after the metadata, but before the message"
+            ),
+        )
+
+        config_builder.add(
+            key=STRUCTLOG_LOGGING_FORMAT_KEY,
+            default_value=",".join(list(STRUCTLOG_LOGGING_METADATA_DEFAULT.keys())),
+            description=(f"Possible values: json, console"),
+        )
+
+    @hookimpl
+    def vdk_initialize(self, context: CoreContext):
+        metadata_keys = context.configuration.get_value(STRUCTLOG_LOGGING_METADATA_KEY)
+        logging_formatter = context.configuration.get_value(
+            STRUCTLOG_LOGGING_FORMAT_KEY
+        )
+
+        formatter, metadata_filter = create_formatter(logging_formatter, metadata_keys)
+        handler = logging.StreamHandler(sys.stderr)
+        handler.addFilter(metadata_filter)
+        handler.setFormatter(formatter)
+        root_logger = logging.getLogger()
+        root_logger.removeHandler(root_logger.handlers[0])
+        root_logger.addHandler(handler)
+
+    @hookimpl(hookwrapper=True)
+    def initialize_job(self, context: JobContext) -> None:
+        logging_formatter = context.core_context.configuration.get_value(
+            STRUCTLOG_LOGGING_FORMAT_KEY
+        )
+        metadata_keys = context.core_context.configuration.get_value(
+            STRUCTLOG_LOGGING_METADATA_KEY
+        )
+
+        formatter, metadata_filter = create_formatter(logging_formatter, metadata_keys)
+        job_name_adder = AttributeAdder("vdk_job_name", context.name)
+
+        root_logger = logging.getLogger()
+        root_logger.removeHandler(root_logger.handlers[0])
+        handler = logging.StreamHandler(sys.stderr)
+        handler.setFormatter(formatter)
+        handler.addFilter(metadata_filter)
+        handler.addFilter(job_name_adder)
+        root_logger.addHandler(handler)
```

Comment:
this is duplicated, maybe split it off into a separate function?

Reply:
I see your point, but we should do this after we move all the log configuration logic inside this plugin. Right now, we have to remove the default handler in some places and it's not ideal to put this in a separate method.